File: src/main/java/com/conference/website/integration/FakeBuzzClient.java
package com.conference.website.integration;

import org.springframework.stereotype.Component;
import reactor.core.publisher.Mono;

import java.time.Duration;

@Component
public class FakeBuzzClient implements BuzzClient {

    private static final Duration LATENCY = Duration.ofMillis(55);

    @Override
    public Mono<Double> getBuzzScore(Long talkId) {
        if (talkId % 13 == 0) {
            return Mono.delay(LATENCY)
                    .flatMap(ignored -> Mono.error(new IllegalStateException("Buzz stream unavailable")));
        }

        double score = ((talkId % 10) + 1) / 10.0;
        return Mono.just(score)
                .delayElement(LATENCY);
    }
}

File: src/main/kotlin/com/conference/website/integration/FakeBuzzClient.kt
package com.conference.website.integration

import org.springframework.stereotype.Component
import reactor.core.publisher.Mono
import java.time.Duration

@Component
class FakeBuzzClient : BuzzClient {

    override fun getBuzzScore(talkId: Long): Mono<Double> {
        if (talkId % 13 == 0L) {
            return Mono.delay(LATENCY)
                .flatMap { Mono.error<Double>(IllegalStateException("Buzz stream unavailable")) }
        }

        val score = ((talkId % 10) + 1) / 10.0
        return Mono.just(score)
            .delayElement(LATENCY)
    }

    companion object {
        private val LATENCY: Duration = Duration.ofMillis(55)
    }
}
